### ISSUE TITLE
Error when compiling class with static initializer

### ISSUE THREAD

**Issue by mbernard2** [Microsoft/TypeScript#15654]
**TypeScript Version:**  2.3.2

I recently ran into problems with Singletons when compiled to JS.

**Code**

In a single file:
```ts
class A {
    private static _instance: A = new A;
    
    static get instance(): A {
        return A._instance;
    }
    
    constructor() {
        let b = new B;
    }
}

class B {
    
}
```
`A` uses object `B` whose class is defined later in the same file.

TypeScript doesn't report any errors.

**Expected behavior:**
This should run without any errors.

**Actual behavior:**
The following error occurs: `TypeError: B is not a constructor`

Here is the output:
```javascript
var A = (function () {
    function A() {
        var b = new B;
    }
    Object.defineProperty(A, "instance", {
        get: function () {
            return A._instance;
        },
        enumerable: true,
        configurable: true
    });
    return A;
}());
A._instance = new A;
var B = (function () {
    function B() {
    }
    return B;
}());
```
`A`'s constructor is called before `B` is defined, so there's an error.

**Comment by mhegazy** [comment]
This is a tricky one. our current analysis is done only on the syntactic level, we do not analyze what functions can possibly be called during initializing the static property and then guard against use of undefined symbols.

**Comment by mbernard2** [comment]
I think a simple way to solve this would be by moving the assignment statement (`
A._instance = new A;`) after variable `B` is assigned. Or, from TypeScript's point of view, after class `B` is declared.

In my case A and B are circularly dependent, so even though I move B to the top, an error still occurs.

Moving static assignments at the end would solve that problem.

**Comment by mhegazy** [comment]
The compiler in general does not reorganize your code. The position we take is to report errors and let the user decide what to do with it.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
